### PR TITLE
Week1

### PR DESCRIPTION
**Pull request by merrymeerkat**
[week 1](https://www.cis.upenn.edu/~cis194/spring15/hw/01-intro.pdf) exercises

### CODE REVIEW

File: src/Week1/HW1.hs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+-- Exercise 3 -----------------------------------------
+-- Double every second number in a list starting on the left.
+doubleEveryOther :: [Integer] -> [Integer]
+doubleEveryOther l = doubleHelper l False
+
+doubleHelper :: [Integer] -> Bool -> [Integer]
+doubleHelper l double = case l of
+                          (x:xs) -> if double then 2*x:(doubleHelper xs False) else x:(doubleHelper xs True)
+                          []     -> []
```
Comment (typ3c4t) [inline src/Week1/HW1.hs:44]: there's a cool way you can do this with pattern matching
```haskell
doubleEveryOther = \case
  x:x':xs -> x : 2*x' : doubleEveryOther xs
  ys      -> ys    -- this is the case where the list has either 0 or 1 elements in which case we don't do any doubling

File: src/Week1/HW1.hs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+-- concise :D
+sumDigits :: [Integer] -> Integer
+sumDigits l = sum $ sum <$> toRevDigits <$> l
+-- the idea is: you turn each integer into a list of its digits, then get the sum of each of these lists, and finally sum the subresults together 
+
+-- Exercise 5 -----------------------------------------
+
+-- Validate a credit card number using the above functions.
+luhn :: Integer -> Bool
+luhn num = if (lastDigit $ sumDigits $ doubleEveryOther $ toRevDigits num) == 0 then True else False
```
Comment (typ3c4t) [inline src/Week1/HW1.hs]: this is one of those `Bool` redundancies 
```suggestion
luhn num = (lastDigit $ sumDigits $ doubleEveryOther $ toRevDigits num) == 0
```

File: src/Week1/HW1.hs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+hanoi :: Integer -> Peg -> Peg -> Peg -> [Move]
+hanoi 0 _ _ _ = []
+hanoi n a b c = hanoi (n-1) a c b ++ [(a,b)] ++ hanoi (n-1) c b a
+
+-- or
+hanoi' :: Integer -> Peg -> Peg -> Peg -> [Move]
+hanoi' n a b c = if n == 0 then []
+                else hanoi' (n-1) a c b ++ [(a,b)] ++ hanoi' (n-1) c b a
+
+-- which is better?
```
Comment (typ3c4t) [inline src/Week1/HW1.hs:75]: I think the first is easier to read

File: src/Week1/HW1.hs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+revList :: [a] -> [a]
+revList l = helper l [] 
+  where helper x acc = case x of []      -> acc
+                                 (x':xs) -> helper xs (x':acc)
+
+-- or, more concisely
+revList' :: [a] -> [a]
+revList' l = foldl (\acc x -> x:acc) [] l
+
+-- I also found this version on stackOverflow. I was a bit confused at first as to why foldl can simply take 2 arguments here. Now I think it's because the compiler `knows` that if revList'' takes in a list, its foldl definition must also take in a list, so we can just let the l be implicit (?)
```
Comment (typ3c4t) [inline src/Week1/HW1.hs:32]: in fact it's more general than that; if you have a function 
```haskell
f x = someFunctionCombination x
```
then you can replace it with just
```haskell
f = someFunctionCombination
```
formally, this is called [eta reduction](https://wiki.haskell.org/Eta_conversion).  [Point-free style](https://stackoverflow.com/questions/944446/what-is-point-free-style-in-functional-programming) is a style of writing functions where you avoid ever naming the argument, and some hardcore haskellers take this style to an art form... there's even a web app [pointfree.io](http://pointfree.io/) that automatically searches for a point-free implementation of whatever function you give it!

for example, in your function `lastDigit`, if we had a way to put `n` at the end of the line, we could use this trick; luckily we can turn any binary function (those that take two arguments) as an infix operator via backticks, so this is equivalent to writing
```haskell
lastDigit n = mod n 10
-- using infix style
lastDigit n = n `mod` 10
-- rewriting as function application
lastDigit n = (`mod` 10) n
-- and then with our trick
lastDigit = `mod` 10
```

File: src/Week1/HW1.hs (hunk 2 above)
Comment (typ3c4t) [reply src/Week1/HW1.hs]: if you wanted to be pointfree, you could rewrite it as follows
```haskell
luhn num = (lastDigit $ sumDigits $ doubleEveryOther $ toRevDigits num) == 0
-- treating `== 0` as a function to apply
luhn num = (==0) $ lastDigit $ sumDigits $ doubleEveryOther $ toRevDigits num
-- rewriting in terms of compositions
luhn num = ((==0) . lastDigit . sumDigits . doubleEveryOther . toRevDigits) num
-- eta reduction
luhn = (==0) . lastDigit . sumDigits . doubleEveryOther . toRevDigits
```

to be clear, this is totally unnecessary if not even a bit self indulgent, but I wanted to just show this as a proof of concept of pointfree thinking; and I do happen to be one of those people who indulges in such things 😹

**Comment by typ3c4t** [comment]
looks great :)

File: src/Week1/HW1.hs (hunk 4 above)
Comment (merrymeerkat) [reply src/Week1/HW1.hs:32]: Thanks so much for the comments and the explanation! That's pretty cool :D

**Pull request merged**